tls/s2n_x509_validator.c

```diff
@@ -376,6 +380,33 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_chain(struct s2n_x509_
 
         S2N_ERROR_IF(op_code <= 0, S2N_ERR_CERT_UNTRUSTED);
         validator->state = VALIDATED;
+        /* X509_STORE_CTX_get1_chain() returns a validated cert chain if a previous call to X509_verify_cert() was successful.
+         * X509_STORE_CTX_get0_chain() is a better API because it doesn't return a copy. But it's not available for Openssl 1.0.2.
+        * Therefore, we call this variant and clean it up during s2n_x509_validator_wipe.
+        * See the comments here:
+        * https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get1_chain.html
+        */
+        validator->cert_chain_validated = X509_STORE_CTX_get1_chain(validator->store_ctx);
+    }
+
+    if (validator->skip_cert_validation) {
+        POSIX_GUARD(s2n_dup(&cert_chain_blob, &conn->secure.peer_cert_chain));
+    } else if (validator->state == VALIDATED) {
+        DEFER_CLEANUP(struct s2n_stuffer cert_chain_out_stuffer = { 0 }, s2n_stuffer_free);
+        POSIX_GUARD(s2n_stuffer_growable_alloc(&cert_chain_out_stuffer, cert_chain_len));
+
+        for (size_t cert_idx = 0; cert_idx < sk_X509_num(validator->cert_chain_validated); cert_idx++) {
+            X509 *cert = sk_X509_value(validator->cert_chain_validated, cert_idx);
+            uint8_t *cert_data = NULL;
+            int encoded_data_len = i2d_X509(cert, &cert_data);
+            POSIX_ENSURE_GT(encoded_data_len, 0);
+            POSIX_GUARD(s2n_stuffer_write_uint24(&cert_chain_out_stuffer, encoded_data_len));
+            POSIX_GUARD(s2n_stuffer_write_bytes(&cert_chain_out_stuffer, cert_data, encoded_data_len));
+            OPENSSL_free(cert_data);
+        }
+
+        POSIX_GUARD(s2n_dup(&cert_chain_out_stuffer.blob, &conn->secure.peer_cert_chain));
+        conn->secure.peer_cert_chain.size = s2n_stuffer_data_available(&cert_chain_out_stuffer);
```

Comment:
You need to change the size of cert_chain_out_stuffer.blob before you do the s2n_dup, not the size of peer_cert_chain afterwards. Otherwise, you'll potentially allocate more memory than necessary for peer_cert_chain, which was what you were trying to avoid by doing the s2n_dup in the first place.

If you don't care if too much memory is allocated for peer_cert_chain, then there's no point in doing the s2n_dup. Just let it use the original memory allocated.